h2/src/main/org/h2/expression/IntervalOperation.java

```diff
@@ -0,0 +1,320 @@
+
+        /**
+         * Interval multiplied by numeric.
+         */
+        INTERVAL_MULTIPLY_NUMERIC,
+
+        /**
+         * Interval divided by numeric.
+         */
+        INTERVAL_DIVIDE_NUMERIC,
+
+        /**
+         * Date-time minus date-time.
+         */
+        DATETIME_MINUS_DATETIME
+    }
+
+    private final IntervalOpType opType;
+    private Expression left, right;
+    private int dataType;
+
+    public IntervalOperation(IntervalOpType opType, Expression left, Expression right) {
+        this.opType = opType;
+        this.left = left;
+        this.right = right;
+        int l = left.getType(), r = right.getType();
+        switch (opType) {
+        case INTERVAL_PLUS_INTERVAL:
+        case INTERVAL_MINUS_INTERVAL:
+            dataType = Value.getHigherOrder(l, r);
+            break;
+        case DATETIME_PLUS_INTERVAL:
+        case DATETIME_MINUS_INTERVAL:
+        case INTERVAL_MULTIPLY_NUMERIC:
+        case INTERVAL_DIVIDE_NUMERIC:
+            dataType = l;
+            break;
+        case DATETIME_MINUS_DATETIME:
+            if (l == Value.TIME && r == Value.TIME) {
+                dataType = Value.INTERVAL_HOUR_TO_SECOND;
+            } else if (l == Value.DATE && r == Value.DATE) {
+                dataType = Value.INTERVAL_DAY;
+            } else {
+                dataType = Value.INTERVAL_DAY_TO_SECOND;
+            }
+        }
+    }
+
+    @Override
+    public String getSQL() {
+        return '(' + left.getSQL() + ' ' + getOperationToken() + ' ' + right.getSQL() + ')';
+    }
+
+    private char getOperationToken() {
+        switch (opType) {
+        case INTERVAL_PLUS_INTERVAL:
+        case DATETIME_PLUS_INTERVAL:
+            return '+';
+        case INTERVAL_MINUS_INTERVAL:
+        case DATETIME_MINUS_INTERVAL:
+        case DATETIME_MINUS_DATETIME:
+            return '-';
+        case INTERVAL_MULTIPLY_NUMERIC:
+            return '*';
+        case INTERVAL_DIVIDE_NUMERIC:
+            return '/';
+        default:
+            throw DbException.throwInternalError("opType=" + opType);
+        }
+    }
+
+    @Override
+    public Value getValue(Session session) {
+        Value l = left.getValue(session);
+        Value r = right.getValue(session);
+        if (l == ValueNull.INSTANCE || r == ValueNull.INSTANCE) {
+            return ValueNull.INSTANCE;
+        }
+        int lType = l.getType(), rType = r.getType();
+        switch (opType) {
+        case INTERVAL_PLUS_INTERVAL:
+        case INTERVAL_MINUS_INTERVAL: {
+            BigInteger a1 = DateTimeUtils.intervalToAbsolute((ValueInterval) l);
+            BigInteger a2 = DateTimeUtils.intervalToAbsolute((ValueInterval) r);
+            return DateTimeUtils.intervalFromAbsolute(
+                    IntervalQualifier.valueOf(Value.getHigherOrder(lType, rType) - Value.INTERVAL_YEAR),
+                    opType == IntervalOpType.INTERVAL_PLUS_INTERVAL ? a1.add(a2) : a1.subtract(a2));
```

Comment:
NANOS_PER_DAY overflow check?

Reply:
Intervals don't have nanos of day and both `leading` and `remaining` fields are combined into one `BigInteger` in all such places.